File: app-ios/VelaWallet/VelaWallet/Features/Send/FeeExecutor.swift
//
//  FeeExecutor.swift
//  VelaWallet
//
//  The only place the `fee_policy` core touches the outside world.
//
//  Seven operations, every one of them a read. Ported from
//  `app-android/.../feature/send/core/FeeExecutor.kt` (spec 043), which is the
//  desktop's `executor/fee.rs`.
//
//  ## Nothing here is a number this shell chose
//
//  A `nil` from any of these is a FACT the core degrades on — only a failed
//  `eth_gasPrice` triggers its five-gwei default, and it is the core that
//  decides that, not this file. A shell that substituted its own figure would
//  be a second opinion nobody diffed, and the figure in question is what
//  somebody pays.
//

import Foundation
import VelaCore

@MainActor
final class FeeExecutor {

    /// Every operation this executor is required to handle.
    static let operations = [
        "fetch_gas_price",
        "fetch_bundler_quote",
        "fetch_in_band_quotes",
        "fetch_fee_recipient",
        "estimate_user_op_gas",
        "measure_inner_calls",
        "start_ttl",
    ]

    /// `eth_estimateGas({from, to, value, data})` for one call on one chain —
    /// the gas hex, or `nil` when nobody answered. The same read the spine's
    /// `measureCall` makes at submit.
    typealias MeasureCall = (_ chainId: Int, _ from: String, _ to: String, _ valueHex: String, _ data: String) async -> String?

    /// The pool's `eth_estimateGas`, as both the spine and this executor read it.
    static func measuring(with pool: RpcPool) -> MeasureCall {
        { chainId, from, to, valueHex, data in
            let outcome = await pool.call(
                chainId: chainId, method: "eth_estimateGas",
                params: [["from": from, "to": to, "value": valueHex, "data": data]]
            )
            guard case .ok(let value) = outcome, let hex = value as? String, hex.hasPrefix("0x") else { return nil }
            return hex
        }
    }

    private let relay: RelayClient
    /// How a draft is assembled for the estimate. The spine owns the order;
    /// this only needs the JSON the bundler is asked about.
    private let accounts: UserOpSpine.AccountPort
    /// The inner calls' own gas (issue #262 follow-up): the quote measures what
    /// the submit measures, so the fee is priced on the `callGasLimit` the op
    /// will carry. Unwired = nothing measured = the relay's figure.
    private let measureCall: MeasureCall

    init(
        relay: RelayClient,
        accounts: UserOpSpine.AccountPort,
        measureCall: @escaping MeasureCall = { _, _, _, _, _ in nil }
    ) {
        self.relay = relay
        self.accounts = accounts
        self.measureCall = measureCall
    }

    func perform(_ operation: [String: Any]) async -> String {
        let chainId = (operation["chain_id"] as? NSNumber)?.intValue ?? 0
        switch operation["type"] as? String ?? "" {

        case "fetch_gas_price":
            // `want_tip` is false on Tempo: `eth_maxPriorityFeePerGas` is
            // meaningless there and asking corrupts the stablecoin
            // reimbursement. The core decides; this obeys.
            let signals = await relay.gasSignals(
                chainId: chainId, wantTip: operation["want_tip"] as? Bool ?? true
            )
            return CoreJSON.string([
                "type": "gas_price",
                "eth_gas_price": signals.ethGasPrice.map { $0 as Any } ?? NSNull(),
                "base_fee": signals.baseFee.map { $0 as Any } ?? NSNull(),
                "priority_fee": signals.priorityFee.map { $0 as Any } ?? NSNull(),
            ])

        case "fetch_bundler_quote":
            let tier = operation["tier"] as? String ?? "fast"
            return CoreJSON.string([
                "type": "bundler_quote",
                "quote": await relay.bundlerQuote(chainId: chainId, tier: tier)
                    .map { $0 as Any } ?? NSNull(),
            ])

        case "fetch_in_band_quotes":
            let account = operation["account"] as? String ?? ""
            return CoreJSON.string([
                "type": "in_band_quotes",
                "quotes": await relay.inBandQuotes(chainId: chainId, safe: account)
                    .map { $0 as Any } ?? NSNull(),
            ])

        case "fetch_fee_recipient":
            let account = operation["account"] as? String ?? ""
            return CoreJSON.string([
                "type": "fee_recipient",
                "recipient": await relay.accountInfo(chainId: chainId, safe: account)?
                    .feeRecipient.map { $0 as Any } ?? NSNull(),
            ])

        case "estimate_user_op_gas":
            return await estimate(operation, chainId: chainId)

        case "measure_inner_calls":
            return await measure(operation, chainId: chainId)

        case "start_ttl":
            let ms = (operation["ms"] as? NSNumber)?.doubleValue ?? 0
            try? await Task.sleep(nanoseconds: UInt64(max(0, ms) * 1_000_000))
            return CoreJSON.string(["type": "ttl_elapsed"])

        default:
            // Answered, loudly. An unanswered operation leaves the machine
            // waiting and the confirm screen's slide disabled forever, which
            // looks exactly like a wallet that refuses to send.
            print("[vela-wallet] fee_policy: unhandled operation \(operation["type"] ?? "?")")
            return CoreJSON.string(["type": "user_op_gas", "outcome": ["type": "context_unavailable"]])
        }
    }

    /// Simulate the REAL batch — the person's calls plus the fee leg.
    ///
    /// The draft is built with a dummy signature, which is what makes the
    /// estimate honest: a verification cost measured without one is a cost the
    /// submitted operation will exceed.
    ///
    /// **One simulation per operation** (spec 078): the speed in force and
    /// every tier preview ask this for the same calls at the same instant —
    /// nothing in it depends on speed — so they share one flight and, for the
    /// fee-signals window, one held answer (`RelayClient.sharedSimulation`).
    /// Only an `estimated` outcome is held; a refusal or an unreachable relay
    /// is shared with whoever asked at that moment and asked again after.
    private func estimate(_ operation: [String: Any], chainId: Int) async -> String {
        let account = operation["account"] as? String ?? ""
        let deployed = operation["deployed"] as? Bool ?? false
        let key = Self.operationKey(
            account: account, deployed: deployed,
            calls: operation["calls"] as? [[String: Any]] ?? []
        )
        return await relay.sharedSimulation(
            chainId: chainId,
            key: key,
            cacheable: { $0.contains("\"estimated\"") },
            run: { await self.simulate(operation, chainId: chainId, account: account, deployed: deployed) }
        )
    }

    /// The operation a simulation is about, as one string: the account, the
    /// deployment and the calls byte for byte, keys sorted so the same
    /// operation always spells the same key.
    static func operationKey(account: String, deployed: Bool, calls: [[String: Any]]) -> String {
        let object: [String: Any] = [
            "account": account.lowercased(), "deployed": deployed, "calls": calls,
        ]
        guard let data = try? JSONSerialization.data(withJSONObject: object, options: [.sortedKeys])
        else { return "\(account.lowercased()):\(deployed):\(calls.count)" }
        return String(decoding: data, as: UTF8.self)
    }

    private func simulate(
        _ operation: [String: Any], chainId: Int, account: String, deployed: Bool
    ) async -> String {
        let calls = (operation["calls"] as? [[String: Any]] ?? []).map { call in
            UserOpCall(
                to: call["to"] as? String ?? "",
                value: call["value"] as? String ?? "0",
                data: call["data"] as? String ?? "0x"
            )
        }
        let keys = await accounts.keys(of: account)
        guard !keys.isEmpty else {
            // No key means no initCode and no signature shape, so there is
            // nothing to simulate. `context_unavailable` is the core's own word
            // for it (invariant ⑤).
            return CoreJSON.string(["type": "user_op_gas", "outcome": ["type": "context_unavailable"]])
        }
        let nonce = deployed
            ? (await relay.nonce(chainId: chainId, sender: account) ?? "0x0")
            : "0x0"
        let floors = userOpFloors(
            chainId: UInt32(chainId), deployed: deployed, subCalls: UInt32(calls.count + 1)
        )
        guard let draft = try? userOpDraft(
            sender: account, nonce: nonce, deployed: deployed,
            keyHexes: keys.map(\.publicKeyHex), calls: calls,
            fee: .estimateOnly, floors: floors
        ), let json = try? userOpRelayJson(draft: draft, feeToken: nil) else {
            return CoreJSON.string(["type": "user_op_gas", "outcome": ["type": "context_unavailable"]])
        }
        switch await relay.estimateUserOpGas(chainId: chainId, opJson: json) {
        case .estimated(let verification, let call, let preVerification):
            return CoreJSON.string([
                "type": "user_op_gas",
                "outcome": [
                    "type": "estimated",
                    "verification_gas_limit": verification,
                    "call_gas_limit": call,
                    "pre_verification_gas": preVerification,
                ],
            ])
        case .refused:
            // The relay looked at THIS operation and said no. The core reads
            // that as a simulation failure and falls back to its own floors.
            return CoreJSON.string(["type": "user_op_gas", "outcome": ["type": "simulation_failed"]])
        case .unreachable:
            return CoreJSON.string(["type": "user_op_gas", "outcome": ["type": "context_unavailable"]])
        }
    }

    /// Each call on its own, from the Safe's address — one entry per call, in
    /// order, a decimal string or `null`. What an unmeasured call means (no
    /// floor, the relay's figure) is the core's.
    private func measure(_ operation: [String: Any], chainId: Int) async -> String {
        let from = operation["from"] as? String ?? ""
        let calls = operation["calls"] as? [[String: Any]] ?? []
        var gas: [Any] = []
        for call in calls {
            let value = call["value"] as? String ?? "0"
            let data = call["data"] as? String ?? "0x"
            let to = call["to"] as? String ?? ""
            guard let valueHex = UserOpSpine.decimalToHex(value).map({ "0x" + $0 }),
                  let hex = await measureCall(chainId, from, to, valueHex, data),
                  let measured = UInt64(hex.dropFirst(hex.hasPrefix("0x") ? 2 : 0), radix: 16)
            else {
                gas.append(NSNull())
                continue
            }
            gas.append(String(measured))
        }
        return CoreJSON.string(["type": "inner_calls_measured", "gas": gas])
    }

    /// What the core hears when an arm threw: nothing was done.
    static func neutralAnswer(_ operation: [String: Any]) -> String {
        switch operation["type"] as? String ?? "" {
        case "fetch_gas_price":
            return CoreJSON.string([
                "type": "gas_price", "eth_gas_price": NSNull(),
                "base_fee": NSNull(), "priority_fee": NSNull(),
            ])
        case "fetch_bundler_quote":
            return CoreJSON.string(["type": "bundler_quote", "quote": NSNull()])
        case "fetch_in_band_quotes":
            return CoreJSON.string(["type": "in_band_quotes", "quotes": NSNull()])
        case "fetch_fee_recipient":
            return CoreJSON.string(["type": "fee_recipient", "recipient": NSNull()])
        case "measure_inner_calls":
            let count = (operation["calls"] as? [Any])?.count ?? 0
            return CoreJSON.string([
                "type": "inner_calls_measured",
                "gas": Array(repeating: NSNull(), count: count) as [Any],
            ])
        case "start_ttl":
            return CoreJSON.string(["type": "ttl_elapsed"])
        default:
            return CoreJSON.string(["type": "user_op_gas", "outcome": ["type": "context_unavailable"]])
        }
    }
}

File: app-android/vela-wallet/app/src/main/java/app/getvela/wallet/feature/send/core/FeeExecutor.kt
package app.getvela.wallet.feature.send.core

import app.getvela.wallet.core.diagnostics.VelaLog
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.delay
import uniffi.vela_core_uniffi.UserOpCall
import uniffi.vela_core_uniffi.UserOpFeeMode
import uniffi.vela_core_uniffi.userOpDraft
import uniffi.vela_core_uniffi.userOpFloors
import uniffi.vela_core_uniffi.userOpRelayJson

/**
 * The `fee_policy` machine's seven arms (spec 043 T025; `MeasureInnerCalls`
 * since the issue #262 follow-up).
 *
 * Numbers in, numbers out: the three gas signals, the relay's quote, the
 * in-band fee assets, the fee recipient, and a gas estimate for the batch the
 * core built — drafted by the core with its dummy signature so the estimate is
 * byte-identical in shape to what the submit will send. Every padding rule,
 * every price, every "is this quote too high" is the core's.
 */
class FeeExecutor(
    private val relay: RelayClient,
    /** The wallet's founding keys, for an undeployed account's initCode. */
    private val keyHexes: suspend (account: String) -> List<String>,
    /**
     * `eth_estimateGas({from, to, value, data})` for one call — the gas hex, or
     * `null` when nobody answered. The same read `UserOpSpine.measureCall` makes
     * at submit, so the quote prices the `callGasLimit` the op will carry
     * (issue #262 follow-up). Unwired = nothing measured = the relay's figure.
     */
    private val measureCall: suspend (chainId: Int, from: String, to: String, valueHex: String, data: String) -> String? =
        { _, _, _, _, _ -> null },
) {

    suspend fun perform(operation: FeeOperation): FeeShellResult = arm(operation).also { result ->
        VelaLog.event("send.fee.arm", operation::class.simpleName ?: "?", "answer" to summary(result))
    }

    private fun summary(result: FeeShellResult): String = when (result) {
        is FeeShellResult.GasPrice -> "gas=${result.eth_gas_price ?: "-"} base=${result.base_fee ?: "-"} tip=${result.priority_fee ?: "-"}"
        is FeeShellResult.BundlerQuote -> "quote=${result.quote?.max_fee_per_gas ?: "-"}"
        is FeeShellResult.InBandQuotes -> "quotes=${result.quotes?.size ?: "-"}"
        is FeeShellResult.FeeRecipient -> "recipient=${result.recipient?.take(10) ?: "-"}"
        is FeeShellResult.UserOpGas -> "gas=${result.outcome::class.simpleName}"
        is FeeShellResult.InnerCallsMeasured -> "inner=${result.gas.joinToString(",") { it ?: "-" }}"
        FeeShellResult.TtlElapsed -> "ttl"
    }

    private suspend fun arm(operation: FeeOperation): FeeShellResult = when (operation) {
        is FeeOperation.FetchGasPrice -> {
            val signals = relay.gasSignals(operation.chain_id, operation.want_tip)
            FeeShellResult.GasPrice(
                eth_gas_price = signals.ethGasPrice,
                base_fee = signals.baseFee,
                priority_fee = signals.priorityFee,
            )
        }

        is FeeOperation.FetchBundlerQuote ->
            FeeShellResult.BundlerQuote(relay.bundlerQuote(operation.chain_id, operation.tier))

        is FeeOperation.FetchInBandQuotes ->
            FeeShellResult.InBandQuotes(relay.inBandQuotes(operation.chain_id, operation.account))

        is FeeOperation.FetchFeeRecipient -> FeeShellResult.FeeRecipient(
            relay.accountInfo(operation.chain_id, operation.account)?.feeRecipient(),
        )

        is FeeOperation.EstimateUserOpGas -> FeeShellResult.UserOpGas(estimate(operation))

        is FeeOperation.MeasureInnerCalls -> FeeShellResult.InnerCallsMeasured(
            // Each call on its own, from the Safe. What an unmeasured call means
            // (no floor — the relay's figure) is the core's.
            operation.calls.map { call ->
                runCatching {
                    val valueHex = "0x" + java.math.BigInteger(call.value.ifBlank { "0" }).toString(16)
                    measureCall(operation.chain_id, operation.from, call.to, valueHex, call.data)
                        ?.let { java.math.BigInteger(it.removePrefix("0x"), 16).toString() }
                }.getOrNull()
            },
        )

        is FeeOperation.StartTtl -> {
            delay(operation.ms.toLong())
            FeeShellResult.TtlElapsed
        }
    }

    /**
     * `simulateUserOpGas`: a truthful dummy operation — the real nonce for a
     * deployed account, the real initCode for an undeployed one — to the
     * relay's estimator. `ContextUnavailable` is the shell unable to build a
     * TRUTHFUL draft; `SimulationFailed` is a truthful draft the relay refused.
     */
    private suspend fun estimate(operation: FeeOperation.EstimateUserOpGas): FeeGasOutcome = coroutineScope {
        val nonce = if (operation.deployed) {
            relay.nonce(operation.chain_id, operation.account) ?: return@coroutineScope FeeGasOutcome.ContextUnavailable
        } else {
            "0x0"
        }
        val keys = if (operation.deployed) emptyList() else keyHexes(operation.account)
        if (!operation.deployed && keys.isEmpty()) return@coroutineScope FeeGasOutcome.ContextUnavailable
        val calls = operation.calls.map { UserOpCall(to = it.to, value = it.value, data = it.data) }
        val floors = userOpFloors(
            chainId = operation.chain_id.toUInt(),
            deployed = operation.deployed,
            subCalls = calls.size.toUInt(),
        )
        val draft = runCatching {
            userOpDraft(
                sender = operation.account,
                nonce = nonce,
                deployed = operation.deployed,
                keyHexes = keys,
                calls = calls,
                fee = UserOpFeeMode.EstimateOnly,
                floors = floors,
            )
        }.getOrElse { error ->
            VelaLog.failure("send.fee.draft", "could not draft the estimate op", error)
            return@coroutineScope FeeGasOutcome.ContextUnavailable
        }
        val json = async { userOpRelayJson(draft, null) }
        when (val answer = relay.estimateUserOpGas(operation.chain_id, json.await())) {
            is RelayClient.EstimateAnswer.Estimated -> FeeGasOutcome.Estimated(
                verification_gas_limit = answer.verificationGasLimit,
                call_gas_limit = answer.callGasLimit,
                pre_verification_gas = answer.preVerificationGas,
            )
            is RelayClient.EstimateAnswer.Refused -> {
                VelaLog.event("send.fee", "estimate refused", "why" to answer.message.take(120))
                FeeGasOutcome.SimulationFailed
            }
            RelayClient.EstimateAnswer.Unreachable -> FeeGasOutcome.SimulationFailed
        }
    }

    /** What the core hears when an arm threw: nothing was read. */
    fun neutralAnswer(operation: FeeOperation): FeeShellResult = when (operation) {
        is FeeOperation.FetchGasPrice -> FeeShellResult.GasPrice()
        is FeeOperation.FetchBundlerQuote -> FeeShellResult.BundlerQuote(null)
        is FeeOperation.FetchInBandQuotes -> FeeShellResult.InBandQuotes(null)
        is FeeOperation.FetchFeeRecipient -> FeeShellResult.FeeRecipient(null)
        is FeeOperation.EstimateUserOpGas -> FeeShellResult.UserOpGas(FeeGasOutcome.ContextUnavailable)
        is FeeOperation.MeasureInnerCalls -> FeeShellResult.InnerCallsMeasured(operation.calls.map { null })
        is FeeOperation.StartTtl -> FeeShellResult.TtlElapsed
    }
}
